### PR TITLE
Introduce `aleph_bft` Hasher trait implementation

### PR DESCRIPTION
**Pull request by prince-chrismc**
## PR Checklist

- [x] I've built the code `cargo build`. For major changes, `cargo build --workspace --release` is recommended.
- [x] I've run the automated unit tests `cargo test`. This executes our automated unit tests.
- [x] I've not broken any existing tests or functionality. In addition to `cargo test`, I've run `cargo clippy`
- [x] I've not introduced any new known security vulnerabilities. I've run `cargo audit`

- [x] I've included a brief description and a good title of the proposed changes and how to test them.
- [x] I've read the [contributing guidelines](https://github.com/pyrsia/.github/blob/main/contributing.md).
- [x] I've associated an [issue](https://github.com/pyrsia/pyrsia/issues) with this Pull Request. 
- [x] I've checked that the code is up-to-date with the `pyrsia/main` branch.
- [x] I've read ["What is a Good PR?"](https://github.com/pyrsia/pyrsia/blob/main/docs/good_pr.md)
- [x] I've assigned this Pull Request to "pyrsia/collaborators"

## Description

For #322

This adds the crate + a basic trait that required with minimal tests

### CODE REVIEW

File: src/blockchain/src/block.rs [hunk 1]
```diff
@@ -181,9 +184,9 @@ mod tests {
         );
         transactions.push(transaction);
         let block_header = Header::new(PartialHeader::new(
-            hash(b""),
+            HashDigest::new(b""),
```
Comment (chb0github) [inline src/blockchain/src/block.rs:187]: Explain to me again why we have a `PartialHeader` when there is only ever one? Which basically means it is the header

File: src/blockchain/src/blockchain.rs [hunk 2]
```diff
@@ -52,8 +47,6 @@ impl Config {
     pub fn new() -> Self {
         Self {
             blockchain_id: BlockchainId::Pyrsia,
-            hash_algorithm: HashAlgorithm::Keccak,
```
Comment (chb0github) [inline src/blockchain/src/blockchain.rs:55]: Why?

File: src/blockchain/src/storage.rs [hunk 3]
```diff
@@ -58,7 +60,7 @@ pub fn read_last_block(path: String) -> (header::HashDigest, u128, header::Addre
 }
 
 // Unformat the genesis block json string
-pub fn parse_genesis_block(line: &String) -> (header::HashDigest, u128, header::Address) {
+pub fn parse_genesis_block(line: &String) -> (HashDigest, u128, header::Address) {
```
Comment (chb0github) [inline src/blockchain/src/storage.rs:63]: We still have a dedicated GenesisBlock?

**Review by chb0github — APPROVED** [review]
Generally, I don't have an issue - but, I did have some curious questions

File: src/blockchain/src/block.rs (hunk 1 above)
Comment (prince-chrismc) [reply src/blockchain/src/block.rs:187]: The idea was to facility calculating some of the generated fields. "pre mature optimization" syndrome.

We decided on a Zoom call to circle back if this becomes a problem

File: src/blockchain/src/blockchain.rs (hunk 2 above)
Comment (prince-chrismc) [reply src/blockchain/src/blockchain.rs:55]: the `multihash` crate tracks this information

File: src/blockchain/src/storage.rs (hunk 3 above)
Comment (prince-chrismc) [reply src/blockchain/src/storage.rs:63]: This required #402 it will slowly be worked on

**Comment by fishseabowl** [comment]
It would be better if you add some description about this PR.

**Review by chb0github — DISMISSED** [review]

File: src/blockchain/src/blockchain.rs [hunk 4]
```diff
@@ -31,12 +31,6 @@ pub enum BlockchainId {
     Pyrsia,
 }
 
-/// Define Supported Hash Algorithm
-#[derive(Serialize, Deserialize, Debug, Clone)]
-pub enum HashAlgorithm {
```
Comment (prince-chrismc) [inline src/blockchain/src/blockchain.rs:36]: handled by `multihash` crate so zapped

File: src/blockchain/src/crypto/hash_algorithm.rs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+   limitations under the License.
+*/
+
+extern crate aleph_bft;
+
+use multihash::{Code, Multihash, MultihashDigest};
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize, Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct HashDigest {
```
Comment (prince-chrismc) [inline src/blockchain/src/crypto/hash_algorithm.rs:23]: This is what we use to calc all the blockchain hashes that are needed, I added https://docs.rs/aleph-bft/latest/aleph_bft/trait.Hasher.html

So Aleph can 

> used for creating identifiers for blocks or units.

as well

**Review by AbhijithGanesh — DISMISSED** [review]

**Review by fishseabowl — APPROVED** [review]
It looks fine to me.

**Review by fishseabowl — DISMISSED** [review]
It looks fine to me

**Review by chb0github — APPROVED** [review]

**Comment by prince-chrismc** [comment]
This has been approved many times but the merge conflicts keep dismissing them.

**Comment by MitaliBo** [comment]
wow did you just merge it without approvals? resolving conflicts still need another round of review.

**Comment by betarelease** [comment]
This a not a good reason to merge any PR `This has been approved many times but the merge conflicts keep dismissing them.`

**Pull request merged**